docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,490 @@
+
+## Context
+
+Historically modules have exposed their functionality to the state machine via the `AppModule` and `AppModuleBasic`
+interfaces which have the following shortcomings:
+* both `AppModule` and `AppModuleBasic` need to be defined which is counter-intuitive
+* apps need to implement the full interfaces, even parts they don't need
+* interface methods depend heavily on unstable third party dependencies, in particular Tendermint
+* legacy required methods have littered these interfaces for far too long
+
+In order to interact with the state machine, modules have needed to do a combination of these things:
+* get store keys from the app
+* call methods on `sdk.Context` which contains more or less the full set of capability available to modules.
+
+By isolating all the state machine functionality into `sdk.Context`, the set of functionalities available to
+modules are tightly coupled to this type. If there are changes to upstream dependencies (such as Tendermint)
+or new functionalities are desired (such as alternate store types), the changes need impact `sdk.Context` and all
+consumers of it (basically all modules). Also, all modules now receive `context.Context` and need to convert these
+to `sdk.Context`'s with a non-ergonomic unwrapping function.
+
+## Decision
+
+The `core` API proposes a set of core APIs that modules can rely on to interact with the state machine and expose their
+functionalities to it that are designed in a principled way such that:
+* tight coupling of dependencies and unrelated functionalities is minimized or eliminated
+* APIs can have long-term stability guarantees
+* the SDK framework is extensible in a safe a straightforward way
+
+The design principles of the core API are as follows:
+* everything that a module wants to interact with in the state machine is a service
+* all services coordinate state via `context.Context` and don't try to recreate the "bag of variables" approach of `sdk.Context`
+* all independent services are isolated in independent packages with minimal APIs and minimal dependencies
+* the core API should be minimalistic and designed for long-term support (LTS)
+* a "runtime" module will implement all the "core services" defined by the core API and can handle all module
+  functionalities exposed by the core `Handler` type
+* other non-core and/or non-LTS services can be exposed by specific versions of runtime modules or other modules 
+following the same design principles, this includes functionality that interacts with specific non-stable versions of
+third party dependencies such as Tendermint
+* the core API doesn't implement *any* functionality, it just defines types
+* go stable API management principles are followed (TODO: link)
+
+Runtime modules which implement the core API are *intentionally* separate from the core API in order to enable more 
+forks of the runtime module than is possible with the SDK's current tightly coupled `BaseApp` design while still
```

Comment:
What could a fork of SDK's default runtime look like in practice?  Is this a code fork or is runtime to be composable so that users of the sdk can add/adjust functionality?  Is there a need for that currently (is this why osmosis forked)?